File: 24_김수빈/10866.c
// 작성일: 2022년 1월 18일
// 작성자: 김수빈
// 프로그램명: 덱
#include <stdio.h>
#include <string.h>
int deque[10000];
int size = 0, start = 0, end = 0;

int empty() {
	if (size == 0) return 1;
	return 0;
}
void push_front(int n) {
	deque[start] = n;
	start = (start - 1) % 10000;
	size++;
}
void push_back(int n) {
	end = (end + 1) % 10000;
	deque[end] = n;
	size++;
}
int pop_front() {
	if (empty()) return -1;
	size--;
	return deque[(++start) % 10000];
}
int pop_back() {
	if (empty()) return -1;
	size--;
	return deque[(end--) % 10000];
}
int front() {
	if (empty()) return -1;
	return deque[start + 1];
}
int back() {
	if (empty()) return -1;
	return deque[end];
}

int main() {
	int num, data;
	char str[11] = { 0 };

	scanf("%d", &num);
	for (int i = 0; i < num; i++) {
		scanf("%s", str);

		if (!strcmp(str, "push_front")) {
			scanf("%d", &data);
			push_front(data);
		}
		else if (!strcmp(str, "push_back")) {
			scanf("%d", &data);
			push_back(data);
		}
		else if (!strcmp(str, "pop_front"))
			printf("%d\n", pop_front());
		else if (!strcmp(str, "pop_back"))
			printf("%d\n", pop_back());
		else if (!strcmp(str, "size"))
			printf("%d\n", size);
		else if (!strcmp(str, "empty"))
			printf("%d\n", empty());
		else if (!strcmp(str, "front"))
			printf("%d\n", front());
		else if (!strcmp(str, "back"))
			printf("%d\n", back());
	}
	return 0;
}

File: 10866.py
import sys

def push_front(x) :
    deque.insert(0, x)

def push_back(x) :
    deque.append(x)

def pop_front() :
    if deque :
        return deque.pop(0)
    else :
        return -1

def pop_back() :
    if deque :
        return deque.pop()
    else :
        return -1

def size() :
    return len(deque)

def empty() :
    if deque :
        return 0
    else :
        return 1

def front() :
    if deque :
        return deque[0]
    else :
        return -1

def back() :
    if deque :
        return deque[-1]
    else :
        return -1

N = int(sys.stdin.readline())
deque = []

for _ in range(N) :
    order = sys.stdin.readline().rstrip().split()
    if order[0] == 'push_front' :
        push_front(order[1])
    elif order[0] == 'push_back' :
        push_back(order[1])
    elif order[0] == 'pop_front' :
        print(pop_front())
    elif order[0] ==  'pop_back' :
        print(pop_back())
    elif order[0] == 'size' :
        print(size())
    elif order[0] == 'empty' :
        print(empty())
    elif order[0] == 'front' :
        print(front())
    elif order[0] == 'back' :
        print(back())